presto-main/src/main/java/com/facebook/presto/security/RefreshingSystemAccessControl.java

```diff
@@ -0,0 +1,323 @@
+import com.facebook.presto.spi.security.Identity;
+import com.facebook.presto.spi.security.Privilege;
+import com.facebook.presto.spi.security.SystemAccessControl;
+import com.google.common.base.Preconditions;
+import io.airlift.log.Logger;
+
+import java.security.Principal;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+public class RefreshingSystemAccessControl
+        implements SystemAccessControl
+{
+    private static final Logger LOG = Logger.get(RefreshingSystemAccessControl.class);
+
+    static final String REFRESH_PERIOD_SEC = "security.refresh-sec";
+    private static final String DEFAULT_REFRESH_PERIOD_SEC = "300"; // 5 mins
+    static final String REFRESH_ENABLED = "security.refresh-enabled";
+    private static final String DEFAULT_REFRESH = "false";
+    private final ScheduledThreadPoolExecutor service;
+
+    private AtomicReference<SystemAccessControl> delegate = new AtomicReference<>();
+
+    /**
+     * Add optional refreshing to the {@link SystemAccessControl} provided by the factory function. The previous
+     * {@link SystemAccessControl} is provided as an {@link Optional} input - null, when there is no previous element
+     * (as with the first time). If refreshing is not enabled, then the factory function is just used to create the
+     * {@link SystemAccessControl} and returned.
+     *
+     * @param config access control configuration from
+     *         {@link com.facebook.presto.spi.security.SystemAccessControlFactory#create(Map)}
+     * @param factory create the next {@link SystemAccessControl}, optionally informed by the previous
+     * @return a {@link SystemAccessControl} that might be able to refresh itself
+     */
+    public static SystemAccessControl optionallyRefresh(Map<String, String> config,
+            Function<Optional<SystemAccessControl>, SystemAccessControl> factory)
+    {
+        if (Boolean.valueOf(config.getOrDefault(REFRESH_ENABLED, DEFAULT_REFRESH))) {
+            long period = Long.valueOf(config.getOrDefault(REFRESH_PERIOD_SEC, DEFAULT_REFRESH_PERIOD_SEC));
+            return new RefreshingSystemAccessControl(factory, period);
+        }
+        return factory.apply(Optional.empty());
+    }
+
+    public RefreshingSystemAccessControl(
+            Function<Optional<SystemAccessControl>, SystemAccessControl> factory, long refreshPeriodSec)
+    {
+        Runnable command = () -> {
+            try {
+                delegate.set(factory.apply(Optional.ofNullable(delegate.get())));
+            }
+            catch (Exception e) {
+                LOG.error("Failed to reload configuration", e);
+            }
+        };
+        // run it the first time to ensure we have an access control
+        command.run();
```

Comment:
if didn't init delegate with this, we should propagate error now

Reply:
Isn't that what the next line,
```
Preconditions.checkState(this.delegate.get() != null, "No initial system control loaded!");
```
is meant to handle? Maybe im misunderstanding your comment